File: iosApp/iosApp/Pages/Map/AlertIcons.swift
//
//  AlertIcons.swift
//  iosApp
//
//  Created by Simon, Emma on 6/5/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import shared
@_spi(Experimental) import MapboxMaps

enum AlertIcons {
    static let alertIconPrefix = "alert-"
    static let alertZoomClosePrefix = "large-"
    static let alertZoomWidePrefix = "small-"

    static let all: [String] =
        [alertZoomClosePrefix, alertZoomWidePrefix].flatMap { zoomPrefix in
            MapStopRoute.allCases.flatMap { routeType in
                StopAlertState.allCases
                    .filter { state in state != StopAlertState.normal }
                    .map { state in
                        "\(alertIconPrefix)\(zoomPrefix)\(routeType.name.lowercased())-\(state.name.lowercased())"
                    }
            }
        }

    // Expression that's true if the specified route index has no service status set
    private static func alertEmpty(_ index: Int) -> MapboxMaps.Exp {
        Exp(.not) { Exp(.has) {
            MapExp.shared.routeAt(index: Int32(index)).toMapbox()
            Exp(.get) { StopFeaturesBuilder.shared.propServiceStatusKey }
        }}
    }

    // Expression that returns the alert status string for the given route index
    private static func alertStatus(_ index: Int) -> MapboxMaps.Exp {
        Exp(.get) {
            MapExp.shared.routeAt(index: Int32(index)).toMapbox()
            Exp(.get) { StopFeaturesBuilder.shared.propServiceStatusKey }
        }
    }

    private static func getAlertIconName(_ zoomPrefix: String, _ index: Int, _ forBus: Bool) -> MapboxMaps.Exp {
        MapExp.shared.busSwitchExp(forBus: forBus, Exp(.switchCase) {
            Exp(.any) {
                // Check if the index is greater than the number of routes at this stop
                Exp(.gte) { index; Exp(.length) { MapExp.shared.routesExp.toMapbox() } }
                // Or if the alert status at this index is empty
                alertEmpty(index)
                // Or if it's normal
                Exp(.eq) { alertStatus(index); StopAlertState.normal.name }
            }
            // If any of the above are true, don't display an alert icon
            ""
            // Otherwise, use the non-normal alert status and route to get its icon name
            Exp(.concat) {
                alertIconPrefix
                zoomPrefix
                Exp(.downcase) { MapExp.shared.routeAt(index: Int32(index)).toMapbox() }
                "-"
                Exp(.downcase) { alertStatus(index) }
            }
        })
    }

    static func getAlertLayerIcon(_ index: Int, forBus: Bool = false) -> Value<ResolvedImage> {
        .expression(Exp(.step) {
            Exp(.zoom)
            getAlertIconName(alertZoomWidePrefix, index, forBus)
            MapDefaults.shared.closeZoomThreshold
            getAlertIconName(alertZoomClosePrefix, index, forBus)
        })
    }
}

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/map/AlertIcons.kt
package com.mbta.tid.mbta_app.map

import com.mbta.tid.mbta_app.map.style.Exp
import com.mbta.tid.mbta_app.map.style.ResolvedImage
import com.mbta.tid.mbta_app.map.style.downcastToResolvedImage
import com.mbta.tid.mbta_app.model.MapStopRoute
import com.mbta.tid.mbta_app.model.StopAlertState

public object AlertIcons {
    internal val alertIconPrefix = "alert-"
    internal val alertZoomClosePrefix = "large-"
    internal val alertZoomWidePrefix = "small-"

    public val all: List<String> =
        listOf(alertZoomClosePrefix, alertZoomWidePrefix).flatMap { zoomPrefix ->
            MapStopRoute.entries.flatMap { routeType ->
                StopAlertState.entries
                    .filter { state ->
                        state !in
                            setOf(
                                StopAlertState.Elevator,
                                StopAlertState.Normal,
                                StopAlertState.AllClear,
                            )
                    }
                    .map { state ->
                        "${alertIconPrefix}${zoomPrefix}${(routeType.name.lowercase())}-${(state.name.lowercase())}"
                    }
            }
        }

    // Expression that's true if the specified route index has no service status set
    private fun alertEmpty(index: Int): Exp<Boolean> {
        return Exp.not(
            Exp.has(MapExp.routeAt(index), Exp.get(StopFeaturesBuilder.propServiceStatusKey))
        )
    }

    // Expression that returns the alert status string for the given route index
    private fun alertStatus(index: Int): Exp<String> {
        return Exp.get(MapExp.routeAt(index), Exp.get(StopFeaturesBuilder.propServiceStatusKey))
    }

    private fun getAlertIconName(zoomPrefix: String, index: Int, forBus: Boolean): Exp<String> {
        return MapExp.busSwitchExp(
            forBus,
            Exp.case(
                Exp.any(
                    // Check if the index is greater than the number of routes at this stop
                    Exp.ge(Exp(index), Exp.length(MapExp.routesExp)),
                    // Or if the alert status at this index is empty
                    alertEmpty(index),
                    // Or if it's normal
                    Exp.eq(alertStatus(index), Exp(StopAlertState.Normal.name)),
                ) to
                    // If any of the above are true, don't display an alert icon
                    Exp(""),
                // Otherwise, use the non-normal alert status and route to get its icon name
                Exp.concat(
                    Exp(alertIconPrefix),
                    Exp(zoomPrefix),
                    Exp.downcase(MapExp.routeAt(index)),
                    Exp("-"),
                    Exp.downcase(alertStatus(index)),
                ),
            ),
        )
    }

    internal fun getAlertLayerIcon(index: Int, forBus: Boolean = false): Exp<ResolvedImage> {
        return Exp.step(
                Exp.zoom(),
                getAlertIconName(alertZoomWidePrefix, index, forBus),
                Exp(MapDefaults.closeZoomThreshold) to
                    getAlertIconName(alertZoomClosePrefix, index, forBus),
            )
            .downcastToResolvedImage()
    }
}
